reference-implementations/icrc25/signer/src/agent.js

```diff
@@ -23,60 +23,41 @@ export async function call(agent, canisterId, methodName, arg) {
-  const currentRequest = request ?? (await agent.createReadStateRequest?.({ paths: [path] }))
-  const state = await agent.readState(canisterId, { paths: [path] }, undefined, currentRequest)
+export async function pollForCert(agent, canisterId, requestId) {
   if (agent.rootKey == null) throw new Error('Agent root key not initialized before polling')
-  const cert = await Certificate.create({
-    certificate: state.certificate,
-    rootKey: agent.rootKey,
-    canisterId: canisterId
-  })
-  const maybeBuf = cert.lookup([...path, new TextEncoder().encode('status')])
-  let status
-  if (typeof maybeBuf === 'undefined') {
-    // Missing requestId means we need to wait
-    status = RequestStatusResponseStatus.Unknown
-  } else {
-    status = new TextDecoder().decode(maybeBuf)
-  }
+  const path = [new TextEncoder().encode('request_status'), requestId]
 
-  // eslint-disable-next-line default-case
-  switch (status) {
-    case RequestStatusResponseStatus.Replied: {
-      return { 
-        response: cert.lookup([...path, 'reply']),
-        certificate: state.certificate
-      }
+  for (;;) {
+    const request = await agent.createReadStateRequest({paths: [path]})
+    const state = await agent.readState(canisterId, {paths: [path]}, undefined, request)
+    const cert = await Certificate.create({
+      certificate: state.certificate,
+      rootKey: agent.rootKey,
+      canisterId: canisterId
+    })
+    const maybeBuf = cert.lookup([...path, new TextEncoder().encode('status')])
+    let status
+    if (typeof maybeBuf === 'undefined') {
+      // Missing requestId means we need to wait
+      status = RequestStatusResponseStatus.Unknown
+    } else {
+      status = new TextDecoder().decode(maybeBuf)
     }
 
-    case RequestStatusResponseStatus.Received:
-    case RequestStatusResponseStatus.Unknown:
-    case RequestStatusResponseStatus.Processing:
-      // Execute the polling strategy, then retry.
-      await strategy(canisterId, requestId, status)
-      return pollForResponse(agent, canisterId, requestId, strategy, currentRequest)
-
-    case RequestStatusResponseStatus.Rejected: {
-      const rejectCode = new Uint8Array(cert.lookup([...path, 'reject_code']))[0]
-      const rejectMessage = new TextDecoder().decode(cert.lookup([...path, 'reject_message']))
-      throw new Error(
-        `Call was rejected:\n` +
-          `  Request ID: ${Buffer.from(requestId).toString('hex')}\n` +
-          `  Reject code: ${rejectCode}\n` +
-          `  Reject text: ${rejectMessage}\n`
-      )
+    switch (status) {
+      // terminal state, end polling
+      case RequestStatusResponseStatus.Replied:
+      case RequestStatusResponseStatus.Rejected:
+      case RequestStatusResponseStatus.Done:
+        return state.certificate
+
+      // continue polling
+      case RequestStatusResponseStatus.Received:
+      case RequestStatusResponseStatus.Unknown:
+      case RequestStatusResponseStatus.Processing:
+        continue;
```

Comment:
I'd keep using the polling strategy here (a default one used directly instead of being a parameter of this function should do fine) to introduce some sensible delays so the network has time to process the canister call, and optimize the amount of unnecessary network calls in result.